npm/nameSpaceController.go

```diff
@@ -0,0 +1,492 @@
+		if err := nsc.syncNameSpace(key); err != nil {
+			// Put the item back on the workqueue to handle any transient errors.
+			nsc.workqueue.AddRateLimited(key)
+			metrics.SendErrorLogAndMetric(util.NSID, "[processNextWorkItem] Error: failed to syncNameSpace %s. Requeuing with err: %v", key, err)
+			return err
+		}
+		// Finally, if no error occurs we Forget this item so it does not
+		// get queued again until another change happens.
+		nsc.workqueue.Forget(obj)
+		log.Logf("Successfully synced '%s'", key)
+		return nil
+	}(obj)
+
+	if err != nil {
+		utilruntime.HandleError(err)
+		return true
+	}
+
+	return true
+}
+
+// syncNameSpace compares the actual state with the desired, and attempts to converge the two.
+func (nsc *nameSpaceController) syncNameSpace(key string) error {
+	// Get the NameSpace resource with this key
+	nsObj, err := nsc.nameSpaceLister.Get(key)
+	// lock to complete events
+	// TODO: Reduce scope of lock later
+	nsc.npMgr.Lock()
+	defer nsc.npMgr.Unlock()
+	if err != nil {
+		if errors.IsNotFound(err) {
+			utilruntime.HandleError(fmt.Errorf("NameSpace '%s' in work queue no longer exists", key))
+			// find the namespace object from a local cache and start cleaning up process (calling cleanUpDeletedPod function)
+			nsKey := util.GetNSNameWithPrefix(key)
+			cachedNs, found := nsc.npMgr.NsMap[nsKey]
+			// if the namespace does not exists, we do not need to clean up process and retry it
+			if !found {
+				return nil
+			}
+
+			// Found the namespace object from NsMap local cache and start cleaning up processes
+			err = nsc.cleanDeletedNamespace(cachedNs.name, cachedNs.LabelsMap)
+			if err != nil {
+				// need to retry this cleaning-up process
+				metrics.SendErrorLogAndMetric(util.NSID, "Error: %v when namespace is not found", err)
+				return fmt.Errorf("Error: %v when namespace is not found\n", err)
+			}
+			return err
+		}
+	}
+
+	if nsObj.DeletionTimestamp != nil || nsObj.DeletionGracePeriodSeconds != nil {
+		return nsc.cleanDeletedNamespace(
+			util.GetNSNameWithPrefix(nsObj.Name),
+			nsObj.Labels,
+		)
+
+	}
+
+	err = nsc.syncUpdateNameSpace(nsObj)
+	// 1. deal with error code and retry this
+	if err != nil {
+		return fmt.Errorf("failed to sync namespace due to  %s", err.Error())
+	}
+
+	return nil
+}
+
+// InitAllNsList syncs all-namespace ipset list.
+func (npMgr *NetworkPolicyManager) InitAllNsList() error {
+	allNs := npMgr.NsMap[util.KubeAllNamespacesFlag]
+	for ns := range npMgr.NsMap {
+		if ns == util.KubeAllNamespacesFlag {
+			continue
+		}
+
+		if err := allNs.IpsMgr.AddToList(util.KubeAllNamespacesFlag, ns); err != nil {
+			metrics.SendErrorLogAndMetric(util.NSID, "[InitAllNsList] Error: failed to add namespace set %s to ipset list %s with err: %v", ns, util.KubeAllNamespacesFlag, err)
+			return err
+		}
+	}
+
+	return nil
+}
+
+// UninitAllNsList cleans all-namespace ipset list.
+func (npMgr *NetworkPolicyManager) UninitAllNsList() error {
```

Comment:
Question: Who use this function? I cannot find caller. If there is caller, we may check lock?